scripts/list_outdated_dependencies.py

```diff
@@ -0,0 +1,272 @@
+    return dependencies
+
+
+def get_optional_deps_pyproject(modified_pyproject: dict[str, Any]) -> list[str]:
+    """Get a list of the optional dependencies from pyproject.toml"""
+
+    dependencies: list[str] = []
+
+    if "optional-dependencies" in modified_pyproject["project"]:
+        for optional_dependency_list in modified_pyproject["project"][
+            "optional-dependencies"
+        ]:
+            dependencies.extend(
+                modified_pyproject["project"]["optional-dependencies"][
+                    optional_dependency_list
+                ]
+            )
+
+    return dependencies
+
+
+def get_modified_pyproject() -> dict[str, Any]:
+    """Get a copy of pyproject.toml with any self-referencing dependencies removed."""
+    with open(PYPROJECT_TOML_PATH, "rb") as pyproject_toml:
+        pyproject = tomli.load(pyproject_toml)
+
+    modified_pyproject = remove_self_dependencies(pyproject)
+    return modified_pyproject
+
+
+def get_deps_dev() -> list[str]:
+    """Get a list of raw dependency strings from requirements-dev.in"""
+    dependencies = []
+    with open(DEV_DEPS_PATH, encoding="utf-8") as dev_deps:
+        lines = dev_deps.readlines()
+
+    for line in lines:
+        if (
+            line.strip().startswith("#")  # skip comments
+            or len(line.strip()) == 0  # skip empty lines
+            or "requirements-dev-common.in" in line  # skip the inclusion line
+        ):
+            continue
+
+        dependencies.append(line)
+
+    return dependencies
+
+
+def get_version_from_pypi(package_name: str, client: httpx.Client) -> str:
+    """Make a call to PyPI to get the version information about `package_name`."""
+    try:
+        response = client.get(f"https://pypi.org/pypi/{package_name}/json")
+        body = response.json()
+        version = body["info"]["version"]
+    except (httpx.RequestError, KeyError):
+        cli.echo_failure(f"Unable to retrieve information for package '{package_name}'")
+        sys.exit(1)
+
+    return version
+
+
+def get_outdated_deps(dependencies: list[str], strip: bool = False) -> list[list[str]]:
+    """Determine which packages have updates available outside of pinned ranges."""
+    outdated: list[list[str]] = []
+    with httpx.Client(timeout=10) as client:
+        for dependency in dependencies:
+            # Convert string into Requirement object so we can reference its parts
+            requirement = Requirement(dependency)
+
+            pypi_version = get_version_from_pypi(requirement.name, client)
+
+            specified = str(requirement.specifier)
+
+            # Strip the specifier symbols from the front of the string if desired
+            if strip:
+                specified = specified.lstrip("<=>!~")
+
+            # append package name, specified version, and latest available version
+            if not requirement.specifier.contains(pypi_version):
+                outdated.append([requirement.name, specified, pypi_version])
```

Comment:
a tuple would be more idiomatic here than a list (maybe even a named tuple)

Reply:
I did initially have a tuple and then mypy gave me a headache when trying to make `print_table()` flexible (should it be script_utils?). I wanted the rows in `print_table` to have arbitrary column counts and this clashed. I'll take another look at it

Reply:
See below